libusb-game-controller/usb_hid.c

```diff
@@ -0,0 +1,230 @@
+#include "ipc.h"
+#include "syscalls.h"
+#include "usb.h"
+#include "usb_backend.h"
+#include "usb_device_drivers.h"
+#include "usb_hid.h"
+#include "utils.h"
+#include "wiimote.h"
+
+static const usb_device_driver_t *usb_device_drivers[] = {
+    // TODO &ds3_usb_device_driver,
+    // TODO &ds4_usb_device_driver,
+    &dr_usb_device_driver,
+};
+
+static usb_input_device_t usb_devices[MAX_FAKE_WIIMOTES] ATTRIBUTE_ALIGN(32);
+
+static inline usb_input_device_t *get_free_usb_device_slot(void)
+{
+    for (int i = 0; i < ARRAY_SIZE(usb_devices); i++) {
+        if (usb_devices[i].usb == NULL)
+            return &usb_devices[i];
+    }
+
+    return NULL;
+}
+
+static inline const usb_device_driver_t *get_usb_device_driver_for(u16 vid, u16 pid)
+{
+    for (int i = 0; i < ARRAY_SIZE(usb_device_drivers); i++) {
+        if (usb_device_drivers[i]->probe(vid, pid))
+            return usb_device_drivers[i];
+    }
+
+    return NULL;
+}
+
+/* API exposed to USB device drivers */
+const ugc_usb_transfer_t *usb_device_driver_issue_ctrl_transfer_async(usb_input_device_t *device,
+                                                                      u8 requesttype, u8 request,
+                                                                      u16 value, u16 index,
+                                                                      void *data, u16 length,
+                                                                      ugc_transfer_cb callback)
+{
+    return _ugc_usb_backend.ctrl_transfer_async(device->usb, requesttype, request, value, index,
+                                                data, length, callback);
+}
+
+const ugc_usb_transfer_t *usb_device_driver_issue_intr_transfer_async(usb_input_device_t *device,
+                                                                      u8 endpoint, void *data,
+                                                                      u16 length,
+                                                                      ugc_transfer_cb callback)
+{
+    return _ugc_usb_backend.intr_transfer_async(device->usb, endpoint, data, length, callback);
+}
+
+static bool timer_cb_wrapper(ugc_usb_device_t *usb_device)
+{
+    usb_input_device_t *device = ugc_input_device_from_usb(usb_device);
+    bool keep = device->driver->timer ? device->driver->timer(device) : false;
+    return keep;
+}
+
+int usb_device_driver_set_timer(usb_input_device_t *device, int time_us, int repeat_time_us)
+{
+    return _ugc_usb_backend.set_timer(device->usb, time_us, repeat_time_us, timer_cb_wrapper);
+}
+
+static int usb_device_ops_resume(void *usrdata, fake_wiimote_t *wiimote)
+{
+    usb_input_device_t *device = usrdata;
+
+    LOG_DEBUG("usb_device_ops_resume\n");
+
+    if (device->suspended) {
+        /* FIXME: Doesn't work properly with DS3.
+         * It doesn't report any data after suspend+resume... */
+#if 0
+		if (usb_hid_v5_suspend_resume(device->host_fd, device->dev_id, 1, 0) != IOS_OK)
```

Comment:
Should we add a method into `struct ugc_usb_backend_t` for suspend/resume?
Something like:
```c
int (*suspend_resume)(ugc_usb_device_t *device, bool resume);
```
resume = true : resume
resume = false: suspend

I'd add it after `get_device_descriptor` and before the transfer functions.